src/lib/Libtpp/tpp_client.c

```diff
@@ -4224,6 +4330,102 @@ leaf_pkt_handler(int tfd, void *data, int len, void *ctx, void *extra)
 	type = *((char *) data);
 	errno = 0;
 
+	if (type == TPP_AUTH_CTX) {
+		tpp_auth_pkt_hdr_t ahdr = {0};
+		void *data_out = NULL;
+		size_t len_out = 0;
+		size_t len_in = 0;
+		void *data_in = NULL;
+		int is_handshake_done = 0;
+		conn_auth_t *authdata = (conn_auth_t *)extra;
+
+		if (authdata == NULL) {
+			snprintf(tpp_get_logbuf(), TPP_LOGBUF_SZ, "tfd=%d, No auth extra found", tfd);
+			tpp_log_func(LOG_CRIT, __func__, tpp_get_logbuf());
+			return -1;
+		}
+
+		memcpy(&ahdr, data, sizeof(tpp_auth_pkt_hdr_t));
+		if (strcmp(ahdr.auth_type, tpp_conf->auth_type) != 0) {
+			snprintf(tpp_get_logbuf(), TPP_LOGBUF_SZ, "tfd=%d, Authentication method mismatch in connection", tfd);
+			tpp_log_func(LOG_CRIT, __func__, tpp_get_logbuf());
+			return -1;
+		}
+
+		len_in = (size_t)len - sizeof(tpp_auth_pkt_hdr_t);
+		data_in = calloc(1, len_in);
+		if (data_in == NULL) {
+			tpp_log_func(LOG_CRIT, __func__, "Out of memory allocating authdata credential");
+			return -1;
+		}
+		memcpy(data_in, (char *)data + sizeof(tpp_auth_pkt_hdr_t), len_in);
+
+		if (pbs_auth_do_handshake(authdata->authctx, data_in, len_in, &data_out, &len_out, &is_handshake_done) != 0) {
+			free(data_in);
+			return -1;
+		}
+
+		if (len_out > 0) {
+			tpp_auth_pkt_hdr_t ahdr = {0};
+			tpp_chunk_t chunks[2] = {{0}};
+			int fd = ((tpp_router_t *) ((tpp_context_t *)ctx)->ptr)->conn_fd;
+
+			ahdr.type = TPP_AUTH_CTX;
+			strcpy(ahdr.auth_type, tpp_conf->auth_type);
+
+			chunks[0].data = &ahdr;
+			chunks[0].len = sizeof(tpp_auth_pkt_hdr_t);
+
+			chunks[1].data = data_out;
+			chunks[1].len = (int)len_out;
+
+			if (tpp_transport_vsend(fd, chunks, 2) != 0) {
+				snprintf(tpp_get_logbuf(), TPP_LOGBUF_SZ, "tpp_transport_vsend failed, err=%d", errno);
+				tpp_log_func(LOG_CRIT, __func__, tpp_get_logbuf());
+				free(data_out);
+				free(data_in);
+				return -1;
+			}
+			free(data_in);
+			free(data_out);
+		}
+
+		if (is_handshake_done == 1) {
+			return leaf_send_ctl_join(tfd, data, ctx);
+		}
+
+		return (len_out == 0 ? -1 : 0);
+
+	} else if (type == TPP_ENCRYPTED_DATA) {
+		conn_auth_t *authdata = (conn_auth_t *)extra;
+		void *data_out = NULL;
+		size_t len_out = 0;
+		char *msgbuf = NULL;
+
+		if (pbs_auth_decrypt_data == NULL) {
+			tpp_log_func(LOG_CRIT, __func__, "External authentication handlers not installed");
+			return -1;
+		}
+
+		if (pbs_auth_decrypt_data(authdata->authctx, (void *)((char *)data + 1), (size_t)len - 1, &data_out, &len_out) != 0) {
+			return -1;
+		}
+
+		if (len_out == 0) {
+			pbs_asprintf(&msgbuf, "invalid decrypted data len: %d, pktlen: %d", len_out, len - 1);
+			tpp_log_func(LOG_CRIT, __func__, msgbuf);
+			free(msgbuf);
+			return -1;
+		}
+
+		free(data);
```

Comment:
Are you sure about this free()? I am surprised it does not cause an error... AFAICS: data = pkt_start + sizeof(int) and pkt_start = conn->scratch.data, which is freed in free_phy_conn(). Am I missing something?

Reply:
@vchlum Thank you! You are right, we don't need this free(). And till now it didn't give any error because we don't have any pre/post pkt_handler to encrypt/decrypt data in the router so router never sends encrypted data to leaf, till now the only leaf->router was sending encrypted data.